### PR TITLE
Introduce desired control annotation & control status 

### PR DESCRIPTION
**Pull request by elanv**
To improve the spec more declarative, operator managed annotation and status.control have been added to replace the existing imperative fields. I applied it to job cancel feature.

Jobs can be canceled with annotation like:
`flinkclusters.flinkoperator.k8s.io/desired-control: job-cancel`

Savepoint can be created:
`flinkclusters.flinkoperator.k8s.io/desired-control: savepoint`

Progress status is recorded in status.control and annotation is deleted upon completion.

### CODE REVIEW

**Comment by functicons** [comment]
Could you add a section in the user guide on how to cancel a running job? Thanks!

**Comment by elanv** [comment]
Thanks for the review!
I have added the guide.

**Comment by functicons** [comment]
Thanks, I'll test it tomorrow.

**Comment by Mrart** [comment]
@elanv  I see your have add  annotation, But don't remove the job spec "cancelRequested" and "savepointGeneration" ?  I think we can't do it together.

**Comment by elanv** [comment]
@Mrart I think so. However, it may be necessary to decide whether to keep those fields in the current version and deprecate them in the next version or deprecate them just from the current version.

**Comment by elanv** [comment]
@functicons I need to make change a little, so I'll commit more later.

**Comment by functicons** [comment]
@elanv sounds good.

**Comment by elanv** [comment]
I've add savepoint control and fixed reconciliation logic.

**Comment by elanv** [comment]
I have committed all.

**Comment by elanv** [comment]
I accidentally closed and opened it.  The work is done.

**Comment by elanv** [comment]
Changed to check whether the savepoint creation was successful or not with the existence of savepointTriggerID.

File: docs/user_guide.md [hunk 1]
```diff
@@ -305,3 +305,61 @@ make deploy
...
     WATCH_NAMESPACE=<namespace-to-watch>
 ```
+
+### Cancel running Flink job
+
+If you want to cancel a running Flink job, attach control annotation to your FlinkCluster's metadata:
+```
+metadata:
+  annotations:
+    flinkclusters.flinkoperator.k8s.io/desired-control: job-cancel
```
Comment (functicons) [inline docs/user_guide.md:315]: Can we add the command line as well: `kubectl annotate flinkclusters  <name> flinkclusters.flinkoperator.k8s.io/desired-control=job-cancel`?

File: docs/user_guide.md [hunk 2]
```diff
@@ -305,3 +305,61 @@ make deploy
...
+    Update Time:     2020-04-03T10:04:50+09:00
+```
+
+### Create savepoint
+
+If you want to trigger savepoint, attach control annotation to your FlinkCluster's metadata:
+```
+metadata:
+  annotations:
+    flinkclusters.flinkoperator.k8s.io/desired-control: savepoint
```
Comment (functicons) [inline docs/user_guide.md:341]: ditto.

File: api/v1beta1/flinkcluster_validate.go [hunk 3]
```diff
@@ -89,6 +94,39 @@ func (v *Validator) ValidateUpdate(old *FlinkCluster, new *FlinkCluster) error {
 	return nil
 }
 
+func (v *Validator) checkControlAnnotations(old *FlinkCluster, new *FlinkCluster) error {
```
Comment (functicons) [inline api/v1beta1/flinkcluster_validate.go:97]: Can we add a unit test for this function?

File: docs/user_guide.md [hunk 4]
```diff
@@ -305,3 +305,61 @@ make deploy
...
+
+Status:
+  Control:
+    Data:            <nil>
+    Name:            job-cancel
+    State:           Succeeded
+    Update Time:     2020-04-03T10:04:50+09:00
+```
+
+### Create savepoint
```
Comment (functicons) [inline docs/user_guide.md:335]: We have a separate doc for managing savepoints, could you add this section there? I might merge the doc to the user guide in the future.

https://github.com/GoogleCloudPlatform/flink-on-k8s-operator/blob/master/docs/savepoints_guide.md

**Comment by functicons** [comment]
Today when I was testing taking savepoint, I misconfigured the savepoints dir, but from the CR status, it is always in `Progressing` status, I ended up figuring out the cause from JM logs. I'm thinking about how we can surface the error in the status, e.g., can we add a field `retries` and increase it for each failure?

**Comment by elanv** [comment]
> Today when I was testing taking savepoint, I misconfigured the savepoints dir, but from the CR status, it is always in `Progressing` status, I ended up figuring out the cause from JM logs. I'm thinking about how we can surface the error in the status, e.g., can we add a field `retries` and increase it for each failure?

Good idea.
I added new commit for retry count when savepoint failed.

**Comment by elanv** [comment]
It is needed to improve on retry-count implementation. I will commit more later.

File: docs/user_guide.md (hunk 1 above)
Comment (elanv) [reply docs/user_guide.md:315]: Done.

File: docs/user_guide.md (hunk 2 above)
Comment (elanv) [reply docs/user_guide.md]: Done.

File: api/v1beta1/flinkcluster_validate.go (hunk 3 above)
Comment (elanv) [reply api/v1beta1/flinkcluster_validate.go:97]: Done.

File: docs/user_guide.md (hunk 4 above)
Comment (elanv) [reply docs/user_guide.md]: I have added to the guide.

**Comment by elanv** [comment]
Complemented two things.
- When the job stops while the savepoint is in progress, the control status changes to failed.
- Displays the number of retry when savepoint fails

**Comment by functicons** [comment]
I tested again. I misconfigured `savepoints_dir` deliberately, then manually added an annotation for savepoint, it failed after 3 retires as expected, then I manually added an annotation for job cancelling, this time it is always in `Progressing` state. I checked the JM log, it says

```
2020-04-06T17:24:39.645Z        INFO    controllers.FlinkCluster        Failed to take savepoint.       {"cluster": "default/flinkcluster-clickcount", "jobID": "401568f79b7408cf947bcfd96573779d"}
2020-04-06T17:24:39.651Z        ERROR   controllers.FlinkCluster        Failed to cancel job    {"cluster": "default/flinkcluster-clickcount", "jobID": "401568f79b7408cf947bcfd96573779d", "error": "java.util.concurrent.CompletionException: java.util.concurrent.CompletionException: org.apache.flink.runtime.checkpoint.CheckpointException: Failed to trigger savepoint.
```

I would be nice if we can add max retries for job cancel as well.

**Comment by elanv** [comment]
I have added it for job cancel and organized.

File: api/v1beta1/flinkcluster_types.go [hunk 5]
```diff
@@ -69,6 +69,21 @@ const (
 	JobRestartPolicyFromSavepointOnFailure = "FromSavepointOnFailure"
 )
 
+// User requested control
+const (
+	// control annotation key
+	ControlDesiredAnnotation = "flinkclusters.flinkoperator.k8s.io/desired-control"
```
Comment (functicons) [inline api/v1beta1/flinkcluster_types.go:75]: I feel it might be to call it `user-control`.

File: api/v1beta1/flinkcluster_types.go [hunk 6]
```diff
@@ -393,6 +408,24 @@ type FlinkClusterComponentsStatus struct {
 	Job *JobStatus `json:"job,omitempty"`
 }
 
+// Control state
+type FlinkClusterControlState struct {
+	// Control name
+	Name string `json:"name"`
+
+	// Control data
+	Data map[string]string `json:"data,omitempty"`
```
Comment (functicons) [inline api/v1beta1/flinkcluster_types.go:417]: This is the detailed info, maybe rename it to `Details`?

File: api/v1beta1/flinkcluster_validate.go [hunk 7]
```diff
@@ -89,6 +94,35 @@ func (v *Validator) ValidateUpdate(old *FlinkCluster, new *FlinkCluster) error {
...
+		case ControlNameSavepoint:
+			if old.Spec.Job == nil {
+				return fmt.Errorf("savepoint is not allowed for session cluster, annotation: %v", ControlDesiredAnnotation)
+			} else if old.Spec.Job.SavepointsDir == nil {
+				return fmt.Errorf("savepoint is not allowed without spec.job.savepointsDir, annotation: %v", ControlDesiredAnnotation)
+			} else if !IsJobActive(old) {
+				return fmt.Errorf("savepoint is not allowed because job is not existing or already stopped, annotation: %v", ControlDesiredAnnotation)
+			}
+		default:
+			return fmt.Errorf("desired control is not valid, annotation key: %v, value: %v", ControlDesiredAnnotation, newDesiredControl)
```
Comment (functicons) [inline api/v1beta1/flinkcluster_validate.go:120]: Add all available controls in the message.

**Comment by elanv** [comment]
Thanks. Renamed and fixed messages.

File: controllers/flinkcluster_util.go [hunk 8]
```diff
@@ -26,11 +26,18 @@ import (
...
-		getJobManagerServiceName(cluster.ObjectMeta.Name),
+	url := fmt.Sprintf(
+		"http://localhost:8001/api/v1/namespaces/%s/services/%s:ui/proxy",
 		cluster.ObjectMeta.Namespace,
-		*cluster.Spec.JobManager.Ports.UI)
+		getJobManagerServiceName(cluster.ObjectMeta.Name))
+
+	return url
+
+	//return fmt.Sprintf(
```
Comment (functicons) [inline controllers/flinkcluster_util.go:36]: Why do you make this change? It seems to be causing problem in my test, the Flink API cannot get ready.

Comment (elanv) [reply controllers/flinkcluster_util.go]: Its my mistake. Sorry.

**Pull request merged**